### PR TITLE
Preserve step's time execution for a non-running task.

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

I'd like to preserve the same user experience for running and non-running task and being able to check the steps time execution from history.

How it looks now:

Screenshot 2023-10-16 at 13 46 19

How it looks after fix:

Screenshot 2023-10-16 at 13 40 47

Reproduce the issue:

* Open any non running task (completed or failed).

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder/package.json [hunk 1]
```diff
@@ -58,6 +58,7 @@
     "@backstage/plugin-catalog-react": "workspace:^",
     "@backstage/plugin-permission-react": "workspace:^",
     "@backstage/plugin-scaffolder-common": "workspace:^",
+    "@backstage/plugin-scaffolder-node": "workspace:^",
```
Comment (freben) [inline plugins/scaffolder/package.json:61]: You can actually not do this. A frontend package shouldn't import a backend (Node) one.

If you need access to these in the frontend too, you may want to migrate them to `@backstage/plugin-scaffolder-common` and deprecate them in `-node`. But I'm not sure if @benjdlambert has opinions about moving them there.

File: plugins/scaffolder/src/api.ts [hunk 2]
```diff
@@ -170,14 +173,62 @@ export class ScaffolderClient implements ScaffolderApi {
...
+    const task = (await taskResponse.json()) as ScaffolderTask;
+    const taskEvents =
+      (await taskEventsResponse.json()) as SerializedTaskEvent[];
+
+    const stepIdToTimestamps = taskEvents
+      .filter(event => event.type === 'log')
+      .reduce((acc, event) => {
+        const stepId = event.body.stepId as string;
+        if (stepId) {
+          acc.set(stepId, [...(acc.get(stepId) ?? []), event.createdAt]);
```
Comment (freben) [inline plugins/scaffolder/src/api.ts:201]: This is horribly inefficient, you keep re-spreading and re-constructing this array a huge number of times. Why don't you just keep track of a map from id to `{ min, max }`?

File: plugins/scaffolder/src/api.ts [hunk 3]
```diff
@@ -170,14 +173,62 @@ export class ScaffolderClient implements ScaffolderApi {
...
+      }, new Map<string, string[]>());
+
+    const toStartedAt = (stepId: string) => {
+      const timestamps = stepIdToTimestamps.get(stepId);
+      return timestamps ? timestamps[0] : undefined;
+    };
+
+    const toEndedAt = (stepId: string) => {
+      const timestamps = stepIdToTimestamps.get(stepId);
+      return timestamps ? timestamps[timestamps.length - 1] : undefined;
```
Comment (freben) [inline plugins/scaffolder/src/api.ts:213]: is this even true? ongoing tasks will land in this too

**Review by freben — COMMENTED** [review]
I think there's more to this. @benjdlambert ?

File: plugins/scaffolder/src/api.ts (hunk 3 above)
Comment (acierto) [reply plugins/scaffolder/src/api.ts]: this logic is used for the initial state of the task, later on the component listens to the events in task event stream. So for the ongoing task it's not applicable.

**Pull request closed without merging**